exercises/react/example.rs

```diff
@@ -0,0 +1,163 @@
+use std::collections::HashMap;
+
+pub type CellID = usize;
+pub type CallbackID = usize;
+
+struct Cell<'a, T: Copy> {
+    value: T,
+    last_value: T,
+    dependents: Vec<CellID>,
+    cell_type: CellType<'a, T>,
+    callbacks_issued: usize,
+    callbacks: HashMap<CallbackID, Box<FnMut(T) -> () + 'a>>,
+}
+
+enum CellType<'a, T: Copy> {
+    Input,
+    Compute(Vec<CellID>, Box<Fn(&[T]) -> T + 'a>),
+}
+
+impl <'a, T: Copy> Cell<'a, T> {
+    fn new(initial: T, cell_type: CellType<'a, T>) -> Self {
+        Cell {
+            value: initial,
+            last_value: initial,
+            dependents: Vec::new(),
+            cell_type: cell_type,
+            callbacks_issued: 0,
+            callbacks: HashMap::new(),
+        }
+    }
+}
+
+pub struct Reactor<'a, T: Copy> {
+    cells: Vec<Cell<'a, T>>,
+}
+
+impl <'a, T: Copy + PartialEq> Reactor<'a, T> {
+    pub fn new() -> Self {
+        Reactor{
+            cells: Vec::new(),
+        }
+    }
+
+    pub fn create_input(&mut self, initial: T) -> CellID {
+        self.cells.push(Cell::new(initial, CellType::Input));
+        self.cells.len() - 1
+    }
+
+    pub fn create_compute<F: Fn(&[T]) -> T + 'a>(&mut self, dependencies: &[CellID], compute_func: F) -> Result<CellID, &'static str> {
+        let new_id = self.cells.len();
+        for &id in dependencies {
+            match self.cells.get_mut(id) {
+                Some(c) => c.dependents.push(new_id),
```

Comment:
Unsafe. If this function gets called with a valid dependency followed by an invalid one, an entry is added to the valid cell's dependents, and then the dependent cell is not created. We have to check all cells' validity first.